memory_util.h

```diff
@@ -0,0 +1,298 @@
+
+    void take_over(void* data, size_type size);
+
+    void append(const BinaryBuffer&);
+
+    template <typename T>
+    void append(const T t);
+
+    void append(const std::string& str);
+
+    template <typename T>
+    void append(const std::vector<T>& vec);
+
+    template <typename T>
+    void append(const T* data, int num_elements);
+
+    void start_reading();
+
+    template<typename T>
+    void read(T& t);
+
+    void read(std::string& str);
+
+    template <typename T>
+    void read(std::vector<T>& vec);
+
+    template<typename T>
+    void read(T* t, int num_elements);
+
+    size_type size () const;
+
+    bool resize (size_type);
+
+    size_type capacity () const;
+
+    bool reserve (size_type);
+
+    size_type offset() const;
+
+    bool empty () const;
+
+    bool eof() const;
+
+    void clear ();
+
+    uchar* data_mutable ();
+
+    const uchar* data () const;
+
+private:
+    void delete_if_own() {
+        if (own_) {
+            delete [] data_;
+        }
+    }
+
+    uchar* data_;
+    size_type size_;
+    size_type capacity_;
+    bool own_;
+    uchar* read_ptr_;
+};
+
+inline BinaryBuffer::~BinaryBuffer() {
+    delete_if_own();
+}
+
+inline BinaryBuffer::BinaryBuffer(size_type size) {
+    data_ = (size != 0 ? new uchar[size] : NULL);
+    size_ = capacity_ = size;
+    own_ = true;
+}
+
+inline BinaryBuffer::BinaryBuffer(size_type size, size_type capacity) {
+    if (size > capacity) {
+        size = capacity;
+    }
+
+    data_ = (capacity != 0 ? new uchar[capacity] : NULL);
+    size_ = size;
+    capacity_ = capacity;
+    own_ = true;
+}
+
+inline BinaryBuffer::BinaryBuffer(const void* data, size_type size) {
+    if (size != 0) {
+        data_ = new uchar[size];
+        std::memcpy(data_, data, size);
+    } else {
+        data_ = NULL;
+    }
+    size_ = capacity_ = size;
+    own_ = true;
+}
+
+inline BinaryBuffer::BinaryBuffer(const BinaryBuffer& other) {
+    if (other.capacity_) {
+        data_ = new uchar[other.capacity_];
+        if (other.size_) {
+            std::memcpy(data_, other.data_, other.size_);
+        }
+    } else {
+        data_ = NULL;
+    }
+    size_ = other.size_;
+    capacity_ = other.capacity_;
+    own_ = true;
+}
+
+inline BinaryBuffer& BinaryBuffer::operator=(const BinaryBuffer& other) {
+    if (&other != this) {
+        if (other.size_ > capacity_) {
+            delete_if_own();
+            data_ = new uchar[other.capacity_];
+            capacity_ = other.capacity_;
+        }
+        std::memcpy(data_, other.data_, other.size_);
+        size_ = other.size_;
+    }
+    return *this;
+}
+
+inline void BinaryBuffer::assign(const void* data, size_type size) {
+    if (size > capacity_) {
+        delete_if_own();
+        data_ = new uchar[size];
+        capacity_ = size;
+        own_ = true;
+    }
+    std::memcpy(data_, data, size);
+    size_ = size;
+}
+
+inline void BinaryBuffer::take_over(void* data, size_type size) {
```

Comment:
How can this let the outside know that data is taken over? What if the outside code tries to delete data?